src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Windows.cs

```diff
@@ -0,0 +1,252 @@
+                        if (_options.IgnoreInaccessible)
+                        {
+                            return IntPtr.Zero;
+                        }
+                        break;
+                    case Interop.Errors.ERROR_FILE_NOT_FOUND:
+                        error = Interop.Errors.ERROR_PATH_NOT_FOUND;
+                        break;
+                }
+
+                throw Win32Marshal.GetExceptionForWin32Error(error, path);
+            }
+
+            return handle;
+        }
+
+        public bool MoveNext()
+        {
+            if (_lastEntryFound)
+                return false;
+
+            bool acquiredLock = false;
+            FileSystemEntry entry = default;
+
+            Monitor.Enter(_lock, ref acquiredLock);
+
+            try
+            {
+                if (_lastEntryFound)
+                    return false;
+
+                do
+                {
+                    FindNextEntry();
+                    if (_lastEntryFound)
+                        return false;
+
+                    // Calling the constructor inside the try block would create a second instance on the stack.
+                    FileSystemEntry.Initialize(ref entry, _entry, _currentPath, _rootDirectory, _originalRootDirectory);
+
+                    // Skip specified attributes
+                    if ((_entry->FileAttributes & _options.AttributesToSkip) != 0)
+                        continue;
+
+                    if ((_entry->FileAttributes & FileAttributes.Directory) != 0)
+                    {
+                        // Subdirectory found
+                        if (PathHelpers.IsDotOrDotDot(_entry->FileName))
+                        {
+                            // "." or "..", don't process unless the option is set
+                            if (!_options.ReturnSpecialDirectories)
+                                continue;
+                        }
+                        else if (_options.RecurseSubdirectories && ShouldRecurseIntoEntry(ref entry))
+                        {
+                            // Recursion is on and the directory was accepted, Queue it
+                            string subDirectory = PathHelpers.CombineNoChecks(_currentPath, _entry->FileName);
+                            IntPtr subDirectoryHandle = CreateDirectoryHandle(subDirectory);
+                            if (subDirectoryHandle != IntPtr.Zero)
+                            {
+                                try
+                                {
+                                    if (_pending == null)
+                                        _pending = new Queue<(IntPtr, string)>();
+                                    _pending.Enqueue((subDirectoryHandle, subDirectory));
+                                }
+                                catch
+                                {
+                                    // Couldn't queue the handle, close it and rethrow
```

Comment:
This would only happen if we were out of memory. Are we consistently trying to avoid leaking resources in this situation?

Reply:
Yes, we're being _extremely_ careful about this as we're not use SafeHandle